### ISSUE TITLE
Budget Manager for LLM(GPT,VertexAI,Falcon,LLama) and others.

### ISSUE THREAD

**Issue by haseeb-heaven** [KillianLucas/open-interpreter#409]
### Is your feature request related to a problem? Please describe.

I am addressing a key issue here. `Open-interpreter` currently lacks transparency when it comes to the costs associated with utilizing various  (LLMs). These models have pricing structures that are often based on the number of tokens processed, typically billed per 1000 tokens. The users are left in the dark about how much they're spending so i guess this is. big issue.

### Describe the solution you'd like

I want code-interpreter to reach out and grab the pricing details for each LLM from a designated source, like a pricing API or a simple configuration file we keep. Then, every time someone uses an LLM during a session or chat, our app will do some quick math based on how many words or tokens were processed. The result? Users get an instant, user-friendly breakdown of what they're spending, session by session and chat by chat.

### Describe alternatives you've considered

One idea was to have someone update the prices manually within code-interpreter every time there's a change. But let's face it, that's asking for trouble—mistakes can happen, and costs might get all mixed up.

### Additional context

[OpenAI Pricing](https://chatgpt.openai.com/pricing)
[Vertex AI Pricing](https://cloud.google.com/vertex-ai/pricing)

LangChain - Keep track of cost using this [LangChain Cost Tracking](https://github.com/langchain-ai/langchain/blob/116cc7998cac563894bdf0037259db767882b55c/libs/langchain/langchain/callbacks/openai_info.py#L7C1-L7C25)

We can use LangChain to get Cost per token for model like done here in this [GPT Engineer - Issue#706](https://github.com/AntonOsika/gpt-engineer/pull/706)

**Comment by TanmayDoesAI** [comment]
I believe a similar thing is being worked on currently by @krrishdholakia and would be available soon, thank you for your suggestion!

**Comment by haseeb-heaven** [comment]
> I believe a similar thing is being worked on currently by @krrishdholakia and would be available soon, thank you for your suggestion!

Thats much needed because GPT-4 usage it too high i cannot track without this feature.

**Comment by krrishdholakia** [comment]
Hey @haseeb-heaven - i think you're idea is awesome and more advanced than my pr (it simply lets you set a max budget). 

We just exposed a model pricing api 
https://docs.litellm.ai/docs/max_tokens_cost


that makes it easy to access the litellm community-maintained model tokens + cost mappings 

https://github.com/BerriAI/litellm/blob/main/model_prices_and_context_window.json


Please let me know if this is helpful in making your PR. If there's any features you feel are missing let me know here or on Discord (https://discord.com/invite/wuPM9dRgDw)

**Comment by haseeb-heaven** [comment]
> Hey @haseeb-heaven - i think you're idea is awesome and more advanced than my pr (it simply lets you set a max budget).
> 
> We just exposed a model pricing api https://docs.litellm.ai/docs/max_tokens_cost
> 
> that makes it easy to access the litellm community-maintained model tokens + cost mappings
> 
> https://github.com/BerriAI/litellm/blob/main/model_prices_and_context_window.json
> 
> Please let me know if this is helpful in making your PR. If there's any features you feel are missing let me know here or on Discord (https://discord.com/invite/wuPM9dRgDw)
> 
> == Re: langchain,
> 
> looks like it's only for OpenAI models - https://github.com/langchain-ai/langchain/blob/116cc7998cac563894bdf0037259db767882b55c/libs/langchain/langchain/callbacks/openai_info.py#L7C1-L7C25?
> 
> and not actively maintained (last commit >1mo. ago).
> 
> Let me know if i'm missing anything.

Yes LiteLLM solved this issues with budget manager, but i have one issue that these prices are hardcoded into API itself and we need to have tracking mechanism which tracks if in future OpenAI or VertexAI or any other LLM's changes their Price then there would be problem.

And while your LiteLLM solved it but we need to add option into `open-interpreter` that user will enter and see its api usage directly or kind of budget manager inside the app itself.

So in short there are two issues.
1.Hard-Coded API Pricing.
2.Bugdet manager option inside `code-interpreter` app itself.

And if there is already PR for these please let me know.

**Comment by haseeb-heaven** [comment]
PR: https://github.com/KillianLucas/open-interpreter/pull/316

**Comment by krrishdholakia** [comment]
what does this mean? 
>  have tracking mechanism which tracks if in future OpenAI or VertexAI or any other LLM's changes their Price 

If you're trying to build this - that's awesome.

**Comment by haseeb-heaven** [comment]
> what does this mean?
> 
> > have tracking mechanism which tracks if in future OpenAI or VertexAI or any other LLM's changes their Price
> 
> If you're trying to build this - that's awesome.

Yes that would be great, i will look for all the possible solutions to do that.

**Comment by vbwyrde** [comment]
I hope I'm not out of place making this suggestion, but since we're talking about predictive models, what would be even nicer, imo, is a predictive cost feature that allows you to write the request, then it goes and gets the current price per token of the model(s) that is being targeted and does a prediction on the cost before you agree to run it.  Some requests, for example, might look small in terms of numbers of tokens, but have a big impact in terms of the cost of resolution.  For example someone could put in "please show me the complete decimal sequence of Pi."   That's a handful of tokens that would, if run to completion be quite costly.  What would be great is for OI to do some quick math and say "Excuse me sir, but are you sure you want to run that?  The estimated cost will be 100,000,000,000,000,000^23 and take 12.9 quintillion years to calculate."   To which you could neatly type "n".  Something like that possible?  It would be super if it could be done.  Might save some folks a bit of heartache.

**Comment by haseeb-heaven** [comment]
> I hope I'm not out of place making this suggestion, but since we're talking about predictive models, what would be even nicer, imo, is a predictive cost feature that allows you to write the request, then it goes and gets the current price per token of the model(s) that is being targeted and does a prediction on the cost before you agree to run it. Some requests, for example, might look small in terms of numbers of tokens, but have a big impact in terms of the cost of resolution. For example someone could put in "please show me the complete decimal sequence of Pi." That's a handful of tokens that would, if run to completion be quite costly. What would be great is for OI to do some quick math and say "Excuse me sir, but are you sure you want to run that? The estimated cost will be $100,000,000,000,000,000^23 and take 12.9 quintillion years to calculate." To which you could neatly type "n". Something like that possible? It would be super if it could be done. Might save some folks a bit of heartache.

You can easily set budget for your API token from OpenAI so no need for in the application i guess because this can be managed from their side but still for checking price which seems higher and costs more can be added to budget manager feature.

**Comment by krrishdholakia** [comment]
@vbwyrde do you mean this? 
https://github.com/BerriAI/litellm/blob/66a3c59ebe8a1b1fd0799f7dbbeafe18601d1903/litellm/budget_manager.py#L62

**Comment by vbwyrde** [comment]
Thanks @krrishdholakia ... If I read that correctly what it does is tell you how much it already cost to run a given prompt after the fact (so it tells you how much you spent).  Is that right?  What I was suggesting is something more predictive... it predicts how much it WILL cost IF you run a given prompt by looking at it and estimating / predicting how much completing the inference will cost.  Again, this may be crazy talk, but that's what I was thinking would be a good feature to have handy.  So if you have a prompt that you suspect WILL be expensive to complete, or worse you don't realize that it will be expensive, then you get notified that says "Estimated Cost to run this prompt: $50", etc.   That way you can elect to not run expensive prompts if you can't afford it.  I think with single prompts given token limits there would be an upper bound to cost per prompt, but with Agents that can chunk work up and repeatedly prompt on their own to get results, the ability to predict the Total Estimated Cost might be useful.  Again, forgive me if I'm not understanding something.  I'm kind of new to all this and may be missing a key piece of information or concept.  Thanks.

**Comment by haseeb-heaven** [comment]
> Thanks @krrishdholakia ... If I read that correctly what it does is tell you how much it already cost to run a given prompt after the fact (so it tells you how much you spent). Is that right? What I was suggesting is something more predictive... it predicts how much it WILL cost IF you run a given prompt by looking at it and estimating / predicting how much completing the inference will cost. Again, this may be crazy talk, but that's what I was thinking would be a good feature to have handy. So if you have a prompt that you suspect WILL be expensive to complete, or worse you don't realize that it will be expensive, then you get notified that says "Estimated Cost to run this prompt: $50", etc. That way you can elect to not run expensive prompts if you can't afford it. I think with single prompts given token limits there would be an upper bound to cost per prompt, but with Agents that can chunk work up and repeatedly prompt on their own to get results, the ability to predict the Total Estimated Cost might be useful. Again, forgive me if I'm not understanding something. I'm kind of new to all this and may be missing a key piece of information or concept. Thanks.

Yes i got your concept of adding future predictions of cost of request which is verg hard to determine because its depends upon lots of various factors first we need to add feature of budget manager into interpreter then only we can improve your requested feature in future.

And people mostly use this for small task like .
1.Can you compress this image.
2.Create folder with some data in it.

 and small tasks like this, 
so i don’t think we need to add predictions for right now but will definitely add this in future

**Comment by vbwyrde** [comment]
Ah, ok, I get ya.  Thanks.  And yes, I was kind of looking further ahead with my suggestion, for when full blow agents come into play.  In those cases, being able to predict costs in advance of requests will be very useful.  Anyway, thanks.  Glad if I can provide any useful ideas for the project.

**Comment by haseeb-heaven** [comment]
> Ah, ok, I get ya. Thanks. And yes, I was kind of looking further ahead with my suggestion, for when full blow agents come into play. In those cases, being able to predict costs in advance of requests will be very useful. Anyway, thanks. Glad if I can provide any useful ideas for the project.

Thanks for your suggestions keep that coming always good to have more ideas to improve product.

**Comment by ericrallen** [comment]
It’s certainly not perfect, and it uses the LiteLLM methods under the hood, but you might want to look at the new `%tokens` magic command.

It will let you see how many tokens have already been used and their estimated cost, as well as estimating the tokens and cost of a prompt you send to the magic command.

This combined with the max budget functionality gives you some more insight into and control over how you’re using tokens.

**Comment by krrishdholakia** [comment]
@ericrallen any suggestions for how we can improve the underlying functionality?

**Comment by ericrallen** [comment]
@krrishdholakia I think we’re all limited by the fact that there aren’t any available API endpoints from OpenAI (or other providers) that expose the current billing information.

I think where you ended up is about the best we can do for the time being.

**Comment by krrishdholakia** [comment]
@ericrallen i didn't see any mention of the openai billing thing on this issue (might've missed it).

Can you say more?

**Comment by ericrallen** [comment]
I just meant that hardcoding the cost per token like LiteLLM is doing and then updating it if/when OpenAI announces price changes is the best we can do without OpenAI providing the model pricing in some sort of API we can hit.

**Comment by getting55** [comment]
> 谢谢 ...如果我没看错的话，它的作用是告诉你在事后运行一个给定的提示已经花了多少钱（所以它会告诉你你花了多少钱）。是吗？我建议的是更具预测性的东西......如果您运行给定的提示并估计/预测完成推理将花费多少，它会预测它的成本。同样，这可能是疯狂的谈话，但这就是我认为将是一个很好的功能。因此，如果您怀疑完成该提示的成本很高，或者更糟糕的是，您没有意识到它会很昂贵，那么您会收到通知，上面写着“运行此提示的估计成本：50 美元”等。这样，如果您负担不起，您可以选择不运行昂贵的提示。我认为，对于给定令牌限制的单个提示，每个提示的成本都会有上限，但是对于可以分块工作并自行重复提示以获得结果的代理，预测总估计成本的能力可能很有用。再次，如果我不明白某些事情，请原谅我。我对这一切有点陌生，可能错过了一个关键的信息或概念。谢谢。

I can offer you discounted amounts, including GPT3.5, 4.0, the latest version of Preview, Vision, if you need to trust my price will make you tempted
<img width="789" alt="屏幕截图 2023-11-09 210622" src="https://github.com/KillianLucas/open-interpreter/assets/73943908/180d2e0f-d615-47c2-8a26-9d8eb9a1a4f0">

**Comment by MikeBirdTech** [comment]
`max_budget` can be set to resolve this. Thanks!
